pkg/cache/stats_counter.go

```diff
@@ -0,0 +1,216 @@
+			Help: "Total number of cache hits.",
+		}),
+		miss: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cache_miss_total",
+			Help: "Total number of cache misses.",
+		}),
+		eviction: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cache_evictions_total",
+			Help: "Total number of cache evictions.",
+		}),
+	}
+	for _, opt := range options {
+		opt(s)
+	}
+	return s
+}
+
+// RecordHits records the number of hits.
+// This method is called by Get and GetIfPresent methods on a cache hit.
+func (c *burrowStatsCounter) RecordHits(hits uint64) {
+	c.hits.Add(float64(hits))
+}
+
+// RecordMisses records the number of misses.
+// This method is called by Get and GetIfPresent methods method on a cache miss.
+func (c *burrowStatsCounter) RecordMisses(miss uint64) {
+	c.miss.Add(float64(miss))
+}
+
+// RecordLoadSuccess records the number of successful loads.
+// This method is called by Get methods on a successful load.
+func (c *burrowStatsCounter) RecordLoadSuccess(loadTime time.Duration) {
+	if !c.trackLoadingCacheStats {
+		return
+	}
+	c.loadSuccess.Inc()
+	c.loadTotalTime.Observe(loadTime.Seconds())
+}
+
+// RecordLoadError records the number of failed loads.
+// This method is called by Get methods on a failed load.
+func (c *burrowStatsCounter) RecordLoadError(loadTime time.Duration) {
+	if !c.trackLoadingCacheStats {
+		return
+	}
+	c.loadError.Inc()
+	c.loadTotalTime.Observe(loadTime.Seconds())
+}
+
+// RecordEviction records the number of evictions.
+func (c *burrowStatsCounter) RecordEviction() {
+	c.eviction.Inc()
+}
+
+// Snapshot records the current stats.
+// This method is called only by Stats method.
+func (c *burrowStatsCounter) Snapshot(s *burrow.Stats) {
+	var err error
+	s.HitCount, err = collectCounter(c.hits)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache hits", "err", err)
+	}
+	s.MissCount, err = collectCounter(c.miss)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache misses", "err", err)
+	}
+	s.EvictionCount, err = collectCounter(c.eviction)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache evictions", "err", err)
+	}
+
+	if !c.trackLoadingCacheStats {
+		return
+	}
+	s.LoadSuccessCount, err = collectCounter(c.loadSuccess)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache load success", "err", err)
+	}
+	s.LoadErrorCount, err = collectCounter(c.loadError)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache load error", "err", err)
+	}
+	totalTime, err := collectHistogramSum(c.loadTotalTime)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache load time", "err", err)
+	}
+	s.TotalLoadTime = time.Duration(totalTime)
+}
+
+func collect(col prometheus.Collector) (*dto.Metric, error) {
+	// This could be done with prometheus.CounterFunc and atomic.Counter easier,
+	// but it would make it harder to use the same code for histograms.
```

Comment:
Why don't we separate it for histograms and counters? I find this code very difficult to understand it doesn't make sense at first glance.

Reply:
Prometheus metrics are optimized for write. I didn't want to keep our own atomic variables and expose them using CounterFuncs. Because they will be mostly used for writing and rarely will be used to get read (except prometheus endpoint).  

Also, I thought having atomic variables for counters and using native histogram metrics for distributing would be more confusing. So I have stuck with native Prometheus metric types. They are not designed to be read, so I had to introduce this collection logic.

Reply:
I wasn't arguing against prometheus metrics, I think that part makes a lot of sense. I meant, even if there is maybe a bit of duplication, I think it would be more easily understandable to read if it was split by histogram and counter types.